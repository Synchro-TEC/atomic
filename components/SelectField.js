import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import nanoid from 'nanoid';

const SelectField = ({
  label,
  disabled,
  options,
  valueKey,
  labelKey,
  onChange,
  hintInfo,
  errorMessage,
  ...otherProps
}) => {
  let labelOpts = {};

  if (hintInfo) {
    labelOpts['data-info'] = hintInfo;
  }

  if (errorMessage) {
    labelOpts['data-error'] = errorMessage;
  }

  const rootClass = cx({ 'is--invalid': errorMessage });

  let selectOptions = {
    onChange: onChange,
    className: rootClass,
  };

  if (disabled) {
    selectOptions.disabled = disabled;
  }

  const selectProps = Object.assign(selectOptions, otherProps);

  const optionsForSelect = options.map(opt => (
    <option key={nanoid()} value={opt[valueKey]}>
      {opt[labelKey]}
    </option>
  ));

  return (
    <label>
      <span {...labelOpts}>{label}</span>
      <div className="sv-select">
        <select {...selectProps}>{optionsForSelect}</select>
        <label>
          <i className="fa fa-angle-down fa-fw" />
        </label>
      </div>
    </label>
  );
};

SelectField.defaultProps = {
  disabled: false,
};

SelectField.propTypes = {
  options: PropTypes.array,
  valueKey: PropTypes.string,
  labelKey: PropTypes.string,
  errorMessage: PropTypes.string,
  disabled: PropTypes.bool,
  hintInfo: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  type: PropTypes.oneOf(['text', 'email', 'password']),
};

export default SelectField;
