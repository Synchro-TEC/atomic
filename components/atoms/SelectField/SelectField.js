import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import nanoid from 'nanoid';

const SelectField = ({
  label,
  disabled,
  options,
  optionValue,
  optionText,
  onChange,
  hintInfo,
  errorMessage,
  ...otherProps
}) => {
  let labelOpts = {};

  if (hintInfo) {
    labelOpts['data-info'] = hintInfo;
  }

  if (errorMessage) {
    labelOpts['data-error'] = errorMessage;
  }

  const rootClass = cx({ 'is--invalid': errorMessage });

  let selectOptions = {
    onChange: onChange,
    className: rootClass,
  };

  if (disabled) {
    selectOptions.disabled = disabled;
  }

  const selectProps = Object.assign(selectOptions, otherProps);

  const optionsForSelect = options.map(opt => (
    <option key={nanoid()} value={typeof optionValue === 'function' ? optionValue(opt) : opt[optionValue]}>
      {typeof optionText === 'function' ? optionText(opt) : opt[optionText]}
    </option>
  ));

  return (
    <label>
      <span {...labelOpts}>{label}</span>
      <div className="sv-select">
        <select {...selectProps}>{optionsForSelect}</select>
        <label>
          <i className="fa fa-angle-down fa-fw" />
        </label>
      </div>
    </label>
  );
};

SelectField.defaultProps = {
  disabled: false,
};

SelectField.propTypes = {
  options: PropTypes.array.isRequired,
  /** Pode ser a key do objeto ou uma function que recebe o obejeto corrente como parâmetro e deve retorna um valor */
  optionValue: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  /** Pode ser a key do objeto ou uma function que recebe o obejeto corrente como parâmetro e deve retorna um valor */
  optionText: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  errorMessage: PropTypes.string,
  disabled: PropTypes.bool,
  hintInfo: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  type: PropTypes.oneOf(['text', 'email', 'password']),
};

export default SelectField;
