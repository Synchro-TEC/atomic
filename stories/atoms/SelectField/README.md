
```jsx
    const optionsOne = [
       { id: 1, name: 'Option A' },
       ...
     ];
     
     const component = () => (
       <form className="sv-form">
         <SelectField
           label="Name"
           errorMessage={text('Name error:', '')}
           hintInfo={text('Hint:', 'Hint Info')}
           onChange={action('on change name')}
           options={optionsOne}
           optionText="name"
           optionValue="id"
           disabled={boolean('Disabled', false)}
         />
         {/* optionText usando uma function */}
         <SelectField
           label="Name"
           optionText={d => `${d.id} - ${d.name}`}
           onChange={action('on change name')}
           options={optionsOne}
           optionValue="id"
         />
       </form>
     );
     
```     
     