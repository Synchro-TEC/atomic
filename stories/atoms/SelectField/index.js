import React from 'react';
import { action } from '@storybook/addon-actions';
import readme from './README.md';
import SelectField from '../../../components/SelectField';
import { boolean, text } from '@storybook/addon-knobs';

const optionsOne = [
  { id: 1, name: 'Option A' },
  { id: 2, name: 'Option B' },
  { id: 3, name: 'Option C' },
  { id: 4, name: 'Option D' },
  { id: 5, name: 'Option E' },
];

const component = () => (
  <form className="sv-form">
    <SelectField
      label="Name"
      errorMessage={text('Name error:', '')}
      hintInfo={text('Hint:', 'Hint Info')}
      onChange={action('on change name')}
      options={optionsOne}
      optionText="name"
      optionValue="id"
      disabled={boolean('Disabled', false)}
    />
    <SelectField
      label="Name"
      optionText={d => `${d.id} - ${d.name}`}
      onChange={action('on change name')}
      options={optionsOne}
      optionValue="id"
    />
  </form>
);

export default [readme, component];
