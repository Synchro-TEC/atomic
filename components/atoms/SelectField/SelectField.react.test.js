import React from 'react';
import SelectField from './SelectField';

describe('Select Component', () => {
  test('Render the select', () => {
    const select = shallow(
      <SelectField options={[{ l: 'k', v: 'v' }, { l: 'k1', v: 'v1' }]} optionValue="v" optionText="l" />
    );
    expect(select).toMatchSnapshot();
  });

  test('Fire change on Select', () => {
    const onChange = jest.fn();
    const selectWrapper = shallow(
      <SelectField
        options={[{ l: 'k', v: 'v' }, { l: 'k1', v: 'v1' }]}
        optionValue="v"
        optionText="l"
        onChange={onChange}
      />
    );
    selectWrapper.find('select').simulate('change', () => {});
    expect(onChange.mock.calls.length).toEqual(1);
  });
});
