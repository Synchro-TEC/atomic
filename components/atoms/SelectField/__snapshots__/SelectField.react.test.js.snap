// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component Render the select 1`] = `
<label>
  <span />
  <div
    className="sv-select"
  >
    <select
      className=""
    >
      <option
        key="4B0wB1ux8iyZeeuKMBupi"
        value="v"
      >
        k
      </option>
      <option
        key="CAPJlWMi7Asag91w8leIN"
        value="v1"
      >
        k1
      </option>
    </select>
    <label>
      <i
        className="fa fa-angle-down fa-fw"
      />
    </label>
  </div>
</label>
`;
